x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/lock_manager_client.ts

```diff
@@ -0,0 +1,375 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { errors } from '@elastic/elasticsearch';
+import { Logger } from '@kbn/logging';
+import { v4 as uuid } from 'uuid';
+import pRetry from 'p-retry';
+import prettyMilliseconds from 'pretty-ms';
+import { once } from 'lodash';
+import { duration } from 'moment';
+import { ElasticsearchClient } from '@kbn/core/server';
+
+export const LOCKS_INDEX_ALIAS = '.kibana_locks';
+export const LOCKS_CONCRETE_INDEX_NAME = `${LOCKS_INDEX_ALIAS}-000001`;
+
+export type LockId = string;
+export interface LockDocument {
+  createdAt: string;
+  expiresAt: string;
+  metadata: Record<string, any>;
+  token: string;
+}
+
+export interface AcquireOptions {
+  /**
+   * Metadata to be stored with the lock. This can be any key-value pair.
+   * This is not mapped and therefore not searchable.
+   */
+  metadata?: Record<string, any>;
+  /**
+   * Time to live (TTL) for the lock in milliseconds.
+   * When a lock expires it can be acquired by another process
+   */
+  ttl?: number;
+}
+
+const createLocksWriteIndexOnce = once(createLocksWriteIndex);
+
+export class LockManager {
+  private token = uuid();
+
+  constructor(
+    private lockId: LockId,
+    private esClient: ElasticsearchClient,
+    private logger: Logger
+  ) {
+    logger.debug(`LockManager initialized with lockId: ${lockId}`);
+  }
+
+  /**
+   * Attempts to acquire a lock by creating a document with the given lockId.
+   * If the lock exists and is expired, it will be released and acquisition retried.
+   */
+  public async acquire({
+    metadata = {},
+    ttl = duration(30, 'minutes').asMilliseconds(),
+  }: AcquireOptions = {}): Promise<boolean> {
+    await createLocksWriteIndexOnce(this.esClient);
+    this.token = uuid();
+    this.logger.debug(
+      `Acquiring lock "${this.lockId}" with ttl = ${prettyMilliseconds(ttl)} and token = ${
+        this.token
+      }`
+    );
+
+    try {
+      const response = await this.esClient.update<LockDocument>({
+        index: LOCKS_CONCRETE_INDEX_NAME,
+        id: this.lockId,
+        scripted_upsert: true,
+        script: {
+          lang: 'painless',
+          source: `
+              // Get the current time on the ES server.
+              long now = System.currentTimeMillis();
+              
+              // If creating the document or if the lock is expired, update it.
+              if (ctx.op == 'create' || Instant.parse(ctx._source.expiresAt).toEpochMilli() < now) {
+                ctx._source.createdAt = Instant.ofEpochMilli(now).toString();
+                ctx._source.expiresAt = Instant.ofEpochMilli(now + params.ttl).toString();
+              } else {
+                ctx.op = 'noop'
+              }
```

Comment:
do you want to call Instant.ofEpochMilli(now) once, and add the ttl manually?

Reply:
Done.